vTPM Support in Cloud-Hypervisor

praveen-pk opened the pull request:
This patchset enables vTPM device that supports CRB Interface with version TPM 2.0.

Sending this version out for review. There are some TODOs in tpm.rs that still need to be addressed. I will address the TODOs while waiting for feedback.

V5:
* drop  dependency on `nix` crate

V4:  
* Add TPM ACPI tables only if tpm parameter is passed at cmdline
 
V3:
* modified the input argument to --tpm following the suggestions below
* Addressed github and CI failures

V2:
* Ran basic checks: cargo fmt, cargo clippy
* Add missing copyright header

likebreath commented:
@praveen-pk Thank you for the contribution. Please fix the errors reported by GH actions and the CI. You may want to switch it to a drafting PR which is a good way to label work-in-progress.

rbradford commented on docs/vtpm.md:
```diff
@@ -0,0 +1,122 @@
+	--cpus boot=4 \
+	--memory size=1024M \
+	--net "tap=,mac=,ip=,mask=" \
+    --vtpm-socket="/var/run/swtpm.socket"
```
For consistency this should be `--vtpm socket=`

sboeuf replied:
I think this should be `--tpm` instead of `--vtpm`.

rbradford commented on src/main.rs:
```diff
@@ -733,6 +742,7 @@ mod unit_tests {
             #[cfg(feature = "gdb")]
             gdb: false,
             platform: None,
+            vtpm_socket: None,
```
I think there should be a new `VtpmConfig` with a single socket member.

sboeuf replied:
Agreed and I would go further saying the structure should be called `TpmConfig`.

praveen-pk replied:
fixed

rbradford commented on vmm/src/acpi.rs:
```diff
@@ -717,6 +729,17 @@ pub fn create_acpi_tables(
         prev_tbl_off = dbg2_offset;
     }
 
+    // TPM2 Table
```
Probably should only create this table if the device is in use.

rbradford commented on vmm/src/device_manager.rs:
```diff
@@ -4145,6 +4181,28 @@ fn numa_node_id_from_memory_zone_id(numa_nodes: &NumaNodes, memory_zone_id: &str
 
     None
 }
+struct VTPMDevice {}
```
nit: whitespace above here.

praveen-pk replied:
fixed

rbradford commented on vtpm/Cargo.toml:
```diff
@@ -0,0 +1,12 @@
+byteorder = "1.4.3"
+libc = "0.2.98"
+log = "0.4.14"
+nix = "0.22.0"
```
nix is a banned dependency. You need to reimplement whatever you need using libc crate or vmm-sys-utils.

praveen-pk replied:
For my context, could you please expand on why nix is banned? Is nix banned under cloud-hypervisor project only or in Rust generally?

FYI, I removed the dependency on nix from vtpm/socket.rs module. Started working on vtpm/tpm_emulator.rs next.

sboeuf replied:
We've always tried to avoid relying on the `nix` crate given it pulls a lot of code into the project, and I think the same idea is applied in the `rust-vmm` crates. Given `nix` is a wrapper around a bunch of syscalls, we either rely on `libc` if the function is provided, or we implement what's missing ourselves.

praveen-pk replied:
Thanks for the details @sboeuf. I dropped dependency on `nix` crate.

rbradford commented on vtpm/src/lib.rs:
```diff
@@ -0,0 +1,6 @@
+#[macro_use]
```
Missing copyright header.

praveen-pk replied:
fixed

rbradford reviewed (commented):
Before updating pushing your PR it would be good to go through and run the steps that the github workflow runs manually.

praveen-pk commented:
> @praveen-pk Thank you for the contribution. Please fix the errors reported by GH actions and the CI. You may want to switch it to a drafting PR which is a good way to label work-in-progress.

Thank you, converted to a drafting PR.

praveen-pk commented:
While working on addressing the errors reported by GitHub Action and CI, I am seeing an interesting issue.

I use ``debug!`` in ``tpm_ioctl.rs`` module to log some debug messages. To import debug macro I added, the following
vtpm/src/lib.rs:

```
extern crate log;
```
This complies fine without any errors and warnings. But ``clippy`` reports the following errors:

```
error: cannot find macro `debug` in this scope
   --> vtpm/src/tpm_ioctl.rs:383:9                                                                 |
383 |         debug!("PtmResult Testing");
    |         ^^^^^
    |
    = note: consider importing this macro:
            log::debug
```

To address this error, I added the following in ``vtpm/src/tpm_ioctl.rs`` :

```
use log::debug;
```
This results in build time warnings:

```
   Compiling vtpm v0.1.0 (/nfs/vtpm_upstream/cloud-hypervisor/vtpm)
warning: unused import: `log::debug`
 --> vtpm/src/tpm_ioctl.rs:8:5
  |
8 | use log::debug;
  |     ^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default
```

Any suggestions on how to address this correctly?

michael2012z commented:
> error: cannot find macro `debug` in this scope
>    --> vtpm/src/tpm_ioctl.rs:383:9                                                                 |
> 383 |         debug!("PtmResult Testing");

The `debug!` is in test code. The printings in unit test may be not very useful, especially when you merely print a test name.

rbradford commented:
You need to do:

```rust
#[macro_use]
extern crate log;
```

In the lib.rs of your crate. You can see this in most of the other crates.

praveen-pk commented:
> > error: cannot find macro `debug` in this scope
> > --> vtpm/src/tpm_ioctl.rs:383:9                                                                 |
> > 383 |         debug!("PtmResult Testing");
> 
> The `debug!` is in test code. The printings in unit test may be not very useful, especially when you merely print a test name.

Thanks for catching this. Seems like macro import into tests were the issue. Resolved after removing the debug! statements from tests.

praveen-pk commented:
Appreciate the feedback. I will address the comments on nix dependency and conditional adding of TPM2 table this week, once I return from my vacation.

rbradford commented on vtpm/src/socket.rs:
```diff
@@ -0,0 +1,192 @@
+        }
+
+        // Send Ancillary data, along with cmds and data
+        let ret = unsafe { libc::sendmsg(self.ctrl_fd, &msghdr, 0) };
```
Try this: `vmm_sys_util::sock_ctrl_msg::ScmSocket` for sending FDs over a unix socket.

sboeuf commented on docs/vtpm.md:
```diff
@@ -0,0 +1,122 @@
+# vTPM
```
I would really try to avoid the `v` in front of TPM given it's implicitly always virtual. The VMM emulates a TPM device, which is why it can be seen as virtual.

sboeuf commented on docs/vtpm.md:
````diff
@@ -0,0 +1,122 @@
+# vTPM
+vtpm in Cloud-Hypervisor is emulated while using ```swtpm``` in the backend.
````
```suggestion
`tpm` in Cloud-Hypervisor is emulated using `swtpm` as the backend.
```

sboeuf commented on docs/vtpm.md:
```diff
@@ -0,0 +1,122 @@
+
+
+## Usage
+`--vtpm-socket`, an optional argument, can be passed to enable vtpm device.
```
I would prefer a more generic option here. A simple `--tpm` is what I would expect, which would accept `socket=` as a parameter. And if someone wants to plug a different type of emulated TPM in the future, we might add a different parameter.

sboeuf commented on vmm/src/device_manager.rs:
```diff
@@ -1975,6 +1984,30 @@ impl DeviceManager {
 
+    fn add_tpm_device(
+        &mut self,
+        _interrupt_manager: &Arc<dyn InterruptManager<GroupConfig = LegacyIrqGroupConfig>>,
```
Why are you passing a reference to the interrupt manager if you're not using it?

sboeuf commented on vmm/src/device_manager.rs:
```diff
@@ -1180,6 +1183,12 @@ impl DeviceManager {
             console_resize_pipe,
         )?;
 
+        let tpm = self.config.lock().unwrap().tpm.clone();
+        if tpm.is_some() {
+            let vtpm = self.add_tpm_device(&legacy_interrupt_manager, tpm.unwrap().path)?;
+            self.bus_devices
+                .push(Arc::clone(&vtpm) as Arc<Mutex<dyn BusDevice>>);
+        }
```
Please replace with the following syntax:
```suggestion
        if let Some(tpm) = self.config.lock().unwrap().tpm.as_ref() {
            let tpm_dev = self.add_tpm_device(&legacy_interrupt_manager, tpm.path)?;
            self.bus_devices
                .push(Arc::clone(&tpm_dev) as Arc<Mutex<dyn BusDevice>>);
        }
```

sboeuf commented on vmm/src/device_manager.rs:
```diff
@@ -4119,6 +4119,29 @@ fn numa_node_id_from_memory_zone_id(numa_nodes: &NumaNodes, memory_zone_id: &str
     None
 }
 
+struct TPMDevice {}
```
Per Rust expectation about structures' names:
```suggestion
struct TpmDevice {}
```

sboeuf commented on vmm/src/vm.rs:
```diff
@@ -2098,6 +2101,7 @@ impl Vm {
             &self.cpu_manager,
             &self.memory_manager,
             &self.numa_nodes,
+            tpm_enabled,
```
```suggestion
            self.config.lock().unwrap().tpm.is_some(),
```

praveen-pk replied:
I couldn't make this change because it results in a cloud-hypervisor being stuck while trying to acquire a Mutex.

Below is the relevant trace from rust-gdb:

```
#0  syscall () at ../sysdeps/unix/sysv/linux/x86_64/syscall.S:38
#1  0x000055f453d75e85 in std::sys::unix::futex::futex_wait () at library/std/src/sys/unix/futex.rs:62
#2  0x000055f45313e308 in std::sys::unix::locks::futex::Mutex::lock_contended () at library/std/src/sys/unix/locks/futex.rs:63
#3  0x000055f453644f60 in std::sys::unix::locks::futex::Mutex::lock (self=0x55f455b37810)
    at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/sys/unix/locks/futex.rs:35
#4  0x000055f45362ecda in std::sys_common::mutex::MovableMutex::raw_lock (self=0x55f455b37810)
    at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/sys_common/mutex.rs:75
#5  0x000055f453341cf5 in std::sync::mutex::Mutex<vmm::vm_config::VmConfig>::lock<vmm::vm_config::VmConfig> (self=0x55f455b37810)
    at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/sync/mutex.rs:266
#6  0x000055f45359530e in vmm::device_manager::{impl#5}::append_aml_bytes (self=0x7f24080027e8, bytes=0x7f240e170f18) at vmm/src/device_manager.rs:4294
#7  0x000055f4535ccdc9 in vmm::acpi::create_dsdt_table (device_manager=0x7f240e17ba80, cpu_manager=0x7f240e17bab8, memory_manager=0x7f240e17bac0) at vmm/src/acpi.rs:183
#8  0x000055f4535cec9f in vmm::acpi::create_acpi_tables (guest_mem=0x7f2408002dc0, device_manager=0x7f240e17ba80, cpu_manager=0x7f240e17bab8,
    memory_manager=0x7f240e17bac0, numa_nodes=0x7f240e17bb10, tpm_enabled=false) at vmm/src/acpi.rs:637
#9  0x000055f453385989 in vmm::vm::Vm::create_acpi_tables (self=0x7f240e17ba68) at vmm/src/vm.rs:2095
#10 0x000055f453385e7a in vmm::vm::Vm::boot (self=0x7f240e17ba68) at vmm/src/vm.rs:2135
#11 0x000055f4531d01bc in vmm::Vmm::vm_boot (self=0x7f240e17ba50) at vmm/src/lib.rs:572
....
```
I expected the parameter to be resolved to a bool value and the lock on `self.config` to be released before `acpi::create_acpi_tables` is invoked. But above invocation results in cloud-hypervisor being stuck with the above stack trace. 

Please let me know if there is a better way to address this issue.

rbradford replied:
It's a deadlock. You were correct to save it to a temporary value first.

sboeuf replied:
Sorry for asking something wrong :-/

praveen-pk replied:
@rbradford why does this result in a deadlock? Doesn't Rust evaluate the parameter before it is passed to the function? I couldn't find any relevant information while googling about this behavior.

rbradford replied:
I know about this form of re-entrancy deadlock from experience. I also couldn't find any documentation on this.

sboeuf commented on vmm/src/vm.rs:
```diff
@@ -2089,7 +2089,10 @@ impl Vm {
             return None;
         }
-
+        let mut tpm_enabled = false;
+        if self.config.lock().unwrap().tpm.is_some() {
+            tpm_enabled = true;
+        }
```
No need for this given my below comment/suggestion.

sboeuf commented on src/main.rs:
```diff
@@ -362,6 +362,14 @@ fn create_app(default_vcpus: String, default_memory: String, default_rng: String
                 .default_value("true"),
+        )
+        .arg(
+            Arg::new("tpm")
```
Ah here I can see it doesn't match the documentation, but it matches my expectations.

sboeuf commented on vmm/src/config.rs:
```diff
@@ -2288,6 +2297,25 @@ impl RestoreConfig {
+
+impl TpmConfig {
+    pub const SYNTAX: &'static str = "TPM device \
+        \"(UNIX Domain Socket from swtpm) path=</path/to/a/file>\"";
```
If we only want to support a UNIX socket, then we should be more specific with the parameter name and call it `socket` instead of `path`. We might want to keep `path` if we plan on relying on a device on the host by providing a path like `/dev/tpm`.
@rbradford wdyt?

praveen-pk replied:
Make sense to have the parameter name as `socket` instead of `path`. I will go ahead and make this change.

rbradford replied:
Socket is good.

praveen-pk replied:
fixed

sboeuf commented on vmm/src/api/openapi/cloud-hypervisor.yaml:
```diff
@@ -573,6 +573,8 @@ components:
           default: false
         platform:
           $ref: "#/components/schemas/PlatformConfig"
+        vtpm_socket:
+          type: string
```
This needs to be updated to match the new syntax.

sboeuf commented on arch/src/x86_64/layout.rs:
```diff
@@ -107,6 +107,11 @@ pub const KVM_TSS_SIZE: u64 = (3 * 4) << 10;
 
+/// TPM Address Space Range
+/// This Address rangs is specific to CRB Interface
+pub const TPM_START: GuestAddress = GuestAddress(0xfed4_0000);
```
So you're saying the spec reserves this address? That's weird for a specification to put such constraint on the implementation, especially given it might not fit every architecture.
I thought the point of TPM2 ACPI table was to advertise where to find the TPM, meaning we wouldn't have to hardcode this value.

sboeuf replied:
Ok well I guess there's no way around it. That being said, please make sure you reserve this specific address range in the allocator so that the allocator won't allow other devices to be positioned on this range.

praveen-pk replied:
will handle this in next revision.

rbradford replied:
@praveen-pk That is the function for reserving PIO addresses not MMIO addresses.

praveen-pk replied:
@rbradford, should I use the following mmio allocator instead?
```
        self.address_manager.allocator
            .lock()
            .unwrap()
            .allocate_platform_mmio_addresses(...)
```

rbradford replied:
I don't know if your address is covered by an allocator so you might not need to do anything but `.allocate_io_addresses()` is definitely wrong.

sboeuf replied:
@praveen-pk you should use the MMIO allocator. And I think this is needed so that we don't end up with the allocator thinking this is a free address range.

praveen-pk replied:
@sboeuf unfortunately, using `allocate_platform_mmio_addresses()` to reserve TPM address range doesn't work.

In my test configuration `platform_mmio_address` range is configured at: `0x3fff_ffef_0000 - 0x3fff_fffe_ffff`. This range is computed at https://github.com/cloud-hypervisor/cloud-hypervisor/blob/6722c303b05c0b3b7f751ec43927db315d4cd52c/vmm/src/memory_manager.rs#L889.

As the TPM address range doesn't fall within the above range, https://github.com/cloud-hypervisor/cloud-hypervisor/blob/6722c303b05c0b3b7f751ec43927db315d4cd52c/vm-allocator/src/address.rs#L93 check fails and cloud-hypervisor exits with error.

```
Error booting VM: VmBoot(DeviceManager(AllocateMmioAddress))
```

MMIO Hole range is `0xC000_0000 - 0xE800_0000`. TPM address range doesn't fall in this range either.

sboeuf replied:
I see, then you don't need to add it to any allocator as there's some kind of grey zone from `0xe800000` to `0x100000000`. I think looking at the IOAPIC is a good example. Given it's located at `0xfec00000`, there's no need to register it to any allocator.

praveen-pk replied:
Gotcha, dropped the allocator step.

sboeuf commented on vmm/src/acpi.rs:
```diff
@@ -264,6 +264,18 @@ fn create_mcfg_table(pci_segments: &[PciSegment]) -> Sdt {
+
+    tpm.write(36, 0_u16); //Platform Class
+    tpm.write(38, 0_u16); // Reserved Space
+    tpm.write(40, 0xfed4_0040_u64); // Address of Control Area
```
Can't we generate the address from our MMIO allocator so that we're not hardcoding all over the codebase?

rbradford commented on tpm/src/lib.rs:
```diff
@@ -7,4 +7,5 @@
 extern crate log;
 
 pub mod socket;
+pub mod tpm_emulator;
```
Do you really need these prefixes? We are already in the tpm crate.

praveen-pk replied:
dropped the `tpm_` prefix

rbradford commented on devices/Cargo.toml:
```diff
@@ -13,12 +13,15 @@ byteorder = "1.4.3"
 vm-memory = "0.9.0"
 vm-migration = { path = "../vm-migration" }
 vmm-sys-util = "0.10.0"
+tpm = { path = "../tpm" }
```
Something has gone wrong with the alphabetical order here.

praveen-pk replied:
this went haywire when I renamed vtpm -> tpm. fixed now.

rbradford replied:
This doesn't look like it got fixed?

praveen-pk replied:
Sorry about this. Looks like I dropped this change accidentally before I pushed. Correctly fixed now.

praveen-pk commented on Cargo.toml:
```diff
@@ -35,6 +35,7 @@ tracer = { path = "tracer" }
 vmm = { path = "vmm" }
 vmm-sys-util = "0.10.0"
 vm-memory = "0.9.0"
+tpm = {path = "tpm"}
```
fixed

praveen-pk commented on tpm/Cargo.toml:
```diff
@@ -0,0 +1,10 @@
+[package]
+name = "tpm"
+edition = "2018"
+version = "0.1.0"
```
fixed

praveen-pk commented on tpm/Cargo.toml:
```diff
@@ -0,0 +1,10 @@
+[package]
+name = "tpm"
+edition = "2018"
```
fixed

praveen-pk commented on tpm/src/lib.rs:
```diff
@@ -0,0 +1,6 @@
+// SPDX-License-Identifier: Apache-2.0
+//
+
+pub mod tpm_ioctl;
```
not really an ioctl. This module contains low/protocol level structs and methods required to send and receive control commands to swtpm.

`protocol.rs` could be a better name but doesn't seem appropriate either.

praveen-pk commented:
The current CI failures are due to the following error while building against musl:
```
error[E0308]: mismatched types
   --> tpm/src/emulator.rs:323:33
    |
323 |             msghdr.msg_iovlen = data_vecs.len();
    |             -----------------   ^^^^^^^^^^^^^^^ expected `i32`, found `usize`
    |             |
    |             expected due to the type of this binding
```
Checking https://github.com/rust-lang/libc/blob/0488a83dc9aba7ec0643ea45d85cc722891e8cc2/src/unix/linux_like/linux/gnu/mod.rs#L95 and https://github.com/rust-lang/libc/blob/0488a83dc9aba7ec0643ea45d85cc722891e8cc2/src/unix/linux_like/linux/musl/b64/mod.rs#L98 I confirmed gnu and musl declare this variable differently.

As expected, if I do cast the value to `i32`, the build fails on `gnu` target, like so:

```
error[E0308]: mismatched types
   --> tpm/src/emulator.rs:323:33
    |
323 |             msghdr.msg_iovlen = data_vecs.len() as i32;
    |             -----------------   ^^^^^^^^^^^^^^^^^^^^^^ expected `usize`, found `i32`
    |             |
    |             expected due to the type of this binding

```

I tried to conditionally assign value to `msghdr.msg_iovlen` with the following code block:

```
            if cfg!(target_env = "musl") {
                msghdr.msg_iovlen = data_vecs.len() as i32; }
            else {
                msghdr.msg_iovlen = data_vecs.len(); }
```
Unfortunately this doesn't work either. Any suggestions on how I can conditionally assign `msghdr.msg_iovlen` based on build target?


Any suggestions on how I can

praveen-pk commented:
To resolve the above issue, I attempted to use attributes like below:

```
            #[cfg(target_env = "musl")]
            msghdr.msg_iovlen = data_vecs.len() as i32;
            #[cfg(not(target_env = "musl"))]
                msghdr.msg_iovlen = data_vecs.len();
```
the build fails claiming attributes on expressions are experimental.

likebreath commented:
> To resolve the above issue, I attempted to use attributes like below:
> 
> ```
>             #[cfg(target_env = "musl")]
>             msghdr.msg_iovlen = data_vecs.len() as i32;
>             #[cfg(not(target_env = "musl"))]
>                 msghdr.msg_iovlen = data_vecs.len();
> ```
> 
> the build fails claiming attributes on expressions are experimental.

@praveen-pk As a workaround, please try: ` msghdr.msg_iovlen = data_vecs.len() as _;`.

Also the inconsistency between rust gnu-libc and musl-libc may worth reporting to the upstream. It is odd to me.

praveen-pk commented:
@likebreath Thanks for the suggestion. That worked!
Reported this issue to upstream at https://github.com/rust-lang/libc/issues/2971.

sboeuf commented:
Now you have the unit tests failing.

rbradford commented:
> 
> @praveen-pk As a workaround, please try: ` msghdr.msg_iovlen = data_vecs.le

Or the other option, which we use when we call `libc::ioctl` is `.try_into().unwrap()`

rbradford commented on tpm/src/lib.rs:
```diff
@@ -0,0 +1,6 @@
+// SPDX-License-Identifier: Apache-2.0
+//
+
+pub mod ioctl;
```
Please don't call it `ioctl` if is not about the syscall.

praveen-pk replied:
I moved all the structs to lib.rs and dropped ioctl.rs. lib.rs seemed to be a better place for these structs and related methods.

rbradford commented on tpm/src/ioctl.rs:
```diff
@@ -0,0 +1,434 @@
+#[derive(Debug, Clone, Copy)]
+pub enum Commands {
+    CmdGetCapability = 1,
+    CmdInit,                // 2
```
Why do you need these numerical comments?

praveen-pk replied:
Just for readability. Dropped these to keep the code clean.

rbradford commented on tpm/src/ioctl.rs:
```diff
@@ -0,0 +1,434 @@
+}
+
+#[derive(Error, Debug)]
+pub enum TPMIocError {
```
There is not need to namespace the enum name Rust gives you namespaces so you can just call it `Error`

rbradford commented on tpm/src/socket.rs:
```diff
@@ -0,0 +1,166 @@
+use vmm_sys_util::sock_ctrl_msg::ScmSocket;
+
+#[derive(Error, Debug)]
+pub enum TPMSocError {
```
Ditto about the enum naming.

praveen-pk replied:
fixed

rbradford commented on tpm/src/socket.rs:
```diff
@@ -0,0 +1,166 @@
+    #[error("Cannot connect to TPM Socket even after retrying for 10 secs")]
+    ConnectToTPMSocket(#[source] anyhow::Error),
+    #[error("Failed to read from TPM socket")]
+    ReadTPMSoc(#[source] anyhow::Error),
```
nit: Inconsistent

praveen-pk replied:
fixed

rbradford commented on tpm/src/socket.rs:
```diff
@@ -0,0 +1,166 @@
+#[derive(Error, Debug)]
+pub enum TPMSocError {
+    #[error("Cannot connect to TPM Socket even after retrying for 10 secs")]
+    ConnectToTPMSocket(#[source] anyhow::Error),
```
You don't need to add the `TPM` here; it's self evident from the context.

praveen-pk replied:
fixed

rbradford commented on tpm/src/socket.rs:
```diff
@@ -0,0 +1,166 @@
+type Result<T> = anyhow::Result<T, TPMSocError>;
+
+#[derive(PartialEq)]
+enum SocDevState {
```
Please don't shorten Socket to Soc

praveen-pk replied:
renamed

rbradford commented on tpm/src/socket.rs:
```diff
@@ -0,0 +1,166 @@
+                "Stream for TPM Socket was not initialized"
+            )));
+        }
+        let mut sock = self.stream.as_ref().unwrap();
```
Because here you shortened it to "sock".

praveen-pk replied:
fixed all the shortened names.

rbradford commented on tpm/src/emulator.rs:
```diff
@@ -0,0 +1,470 @@
+    pub selftest_done: bool,
+}
+
+pub struct TPMEmulator {
```
This commit is the same:

* Don't repeat the context of the types, we know this a TPM emulator as it is in the TPM crate hiearchy.
* Inconsistency with nomenclature (this makes it much harder to search the code)

rbradford commented on devices/src/tpm.rs:
```diff
@@ -0,0 +1,479 @@
+}
+
+/// Return an u8 array from u32
+fn u32_to_u8_arr(data: u32) -> [u8; 4] {
```
Please don't obfuscate the code by wrapping well known Rust behaviour.

praveen-pk replied:
I started thinking, this would be multi line function. Anyway, removed this method as well.

rbradford commented on devices/src/tpm.rs:
```diff
@@ -0,0 +1,479 @@
+    let mut input: [u8; 4] = [0; 4];
+    input.copy_from_slice(&data[0..4]);
+    u32::from_le_bytes(input)
+}
```
Thus function is only called once.

praveen-pk replied:
removed this function

rbradford commented on devices/src/tpm.rs:
```diff
@@ -0,0 +1,479 @@
+            }
+        }
+        debug!(
+            "tpm: MMIO Read: offset {:#X} len {:?} val = {:02X?}  ",
```
You don't need to prefix log messages. The logging infrastructure does that for you.

praveen-pk replied:
fixed all the log messages.

rbradford commented on docs/tpm.md:
```diff
@@ -0,0 +1,122 @@
+	--cpus boot=4 \
+	--memory size=1024M \
+	--net "tap=,mac=,ip=,mask=" \
+    --tpm socket="/var/run/swtpm.socket"
```
nit: Weird indentation?

praveen-pk replied:
fixed

rbradford commented:
@praveen-pk You could undraft this PR and we can give it a final review?

praveen-pk commented:
Addressed all feedback, so far. Undrafted the PR.

rbradford commented on tpm/src/lib.rs:
```diff
@@ -0,0 +1,434 @@
+
+pub trait Ptm {
+    // Get Member Type
+    fn get_memb_type(&self) -> MemberType;
```
nit: Spell out member as you did it in the type name.

praveen-pk replied:
done!

rbradford commented on tpm/src/lib.rs:
```diff
@@ -0,0 +1,434 @@
+    fn update_ptm_with_response(&mut self, buf: &[u8]) -> Result<()>;
+
+    // Update tpm result
+    fn set_res(&mut self, res: u32);
```
nit: Shouldn't this be `set_result_code` for consistency?

praveen-pk replied:
thanks for the catch! updated

rbradford commented on tpm/src/socket.rs:
```diff
@@ -0,0 +1,160 @@
+                }
+                Err(_e) => {}
+            };
+            sleep(Duration::from_millis(1000));
```
What thread does this run in? You can't do this from the vmm thread.

praveen-pk replied:
I dropped the retry logic here. With multiple test confirmed, the retry logic isn't really required.

rbradford commented on tpm/src/socket.rs:
```diff
@@ -0,0 +1,160 @@
+        let size: usize = socket.read(newbuf).map_err(|e| {
+            Error::ReadFromSocket(anyhow!("Failed to read from socket. Error Code {:?}", e))
+        })?;
+        buf[0..size].clone_from_slice(&newbuf[0..size]);
```
If `buf` is larger than the bytes written this will succeed but how will the caller know that fewer bytes were read?

praveen-pk replied:
CRB interface defines the max buf size to TPM_CRB_BUFFER_MAX(3968). I am using a buf of this size to read into. So, I am expecting there will be no case when I will only read a subset of bytes from backend. 

I have also verified all the Control commands implemented in this code. All the responses will fit in this MAX buff size. Hope this addresses your concern.

rbradford commented on tpm/src/socket.rs:
```diff
@@ -0,0 +1,160 @@
+                "Stream for tpm socket was not initialized"
+            )));
+        }
+        let res = match self.state {
```
You could simplify this with `if matches!(self.state,SocketDevState::Connected)`

praveen-pk replied:
updated!

rbradford commented on tpm/src/emulator.rs:
```diff
@@ -0,0 +1,450 @@
+
+        if !emulator.get_established_flag() {
+            emulator.had_startup_error = true;
+        }
```
It would be better to just return an `Err(..)` here if there was failure.

praveen-pk replied:
makes sense. Updated.

rbradford commented on tpm/src/emulator.rs:
```diff
@@ -0,0 +1,450 @@
+    had_startup_error: bool,
+    cmd: Option<BackendCmd>,
+    caps: PtmCap, /* capabilities of the TPM */
+    ctrl_socet: SocketDev,
```
This is a strange shortening.

praveen-pk replied:
This was a typo from when I last expanded the variable name. Fixed it now.

rbradford commented on tpm/src/emulator.rs:
```diff
@@ -0,0 +1,450 @@
+        Ok(())
+    }
+
+    pub fn get_buffer_size(&mut self) -> Result<usize> {
```
This is a strange way to do it. You should set a default buffer size in the constructor. Accessors shouldn't take "&mut self".

praveen-pk replied:
This is defined by protocol. Below is the relevant excerpt from https://github.com/stefanberger/swtpm/blob/master/man/man3/swtpm_ioctls.pod
```
If a 0 is set in the buffer size of the request, the response will return the buffer size that is currently in use.
```

rbradford commented on tpm/src/emulator.rs:
```diff
@@ -0,0 +1,450 @@
+    fn ctrlcmd(
+        &mut self,
+        cmd: Commands,
+        msg: &mut dyn Ptm,
```
Why does this have to be a C style in/out variable. Can't you return separately?

praveen-pk replied:
I couldn't find a strong reason for doing it this way. The original author wrote it this way and I retained it.
If you are strongly against this, I will take some time to refactor this. If not, I would like to merge this version and revisit it later.

rbradford commented on devices/src/tpm.rs:
```diff
@@ -0,0 +1,471 @@
+        tpm.reset()?;
+        Ok(tpm)
+    }
+    fn get_active_locty(&mut self) -> u32 {
```
What is this shortening?

praveen-pk replied:
this name was adopted from Qemu code. Locty here refers to locality. Expanded in code.

rbradford commented on devices/src/tpm.rs:
```diff
@@ -0,0 +1,471 @@
+                }
+                CRB_CTRL_CMD_LADDR => {
+                    self.regs[CRB_CTRL_CMD_LADDR as usize] = v;
+                    return None;
```
This mixture of explicit `return None` and relying on the one outside the `match` is very unreadable.

praveen-pk replied:
Cleaned up all the inner `return None` statements. Only using the `return None` outside the match.

rbradford reviewed (commented):
Because this code can block for upto 10s it can't be in the vmm thread. So you will need to create a new thread to run the code from (and establish appropriate seccomp rules.)

rbradford commented:
> Because this code can block for upto 10s it can't be in the vmm thread. So you will need to create a new thread to run the code from (and establish appropriate seccomp rules.)

Alternatively if you don't need the connection retry/sleeping then it can stay in the vmm thread like vfio-user.

praveen-pk commented:
> > Because this code can block for upto 10s it can't be in the vmm thread. So you will need to create a new thread to run the code from (and establish appropriate seccomp rules.)
> 
> Alternatively if you don't need the connection retry/sleeping then it can stay in the vmm thread like vfio-user.

I am running some tests to confirm if I really need the retry logic. If I don't need the retry logic, I will remove it and leave this method in vmm thread.

rbradford commented on tpm/src/emulator.rs:
```diff
@@ -0,0 +1,452 @@
+pub enum Error {
+    #[error("Input socket path for tpm emulator does not exist")]
+    SocketPathExists(),
+    #[error("Could not initialize emulator's backend")]
```
Missing embedding the error in the string

praveen-pk replied:
fixed

rbradford commented on tpm/src/emulator.rs:
```diff
@@ -0,0 +1,452 @@
+    SocketPathExists(),
+    #[error("Could not initialize emulator's backend")]
+    InitializeEmulator(#[source] anyhow::Error),
+    #[error("Failed to create data fd to pass to swtpm")]
```
Ditto

praveen-pk replied:
fixed

rbradford commented on tpm/src/socket.rs:
```diff
@@ -0,0 +1,139 @@
+// Copyright © 2022, Microsoft Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
```
Why does this file have "OR BSD-3-Clause"?

praveen-pk replied:
Thanks for catching this, it was incorrectly captured. Fixed now.

rbradford reviewed (commented):
Very close now!

praveen-pk commented on tpm/src/emulator.rs:
```diff
@@ -0,0 +1,452 @@
+            if isselftest {
+                let errcode: &[u8; 4] = cmd.output[6..6 + 4]
+                    .try_into()
+                    .expect("slice with incorrect length");
```
addressed

praveen-pk commented on devices/src/tpm.rs:
```diff
@@ -0,0 +1,465 @@
+            set_reg_field(&mut self.regs, CRB_CTRL_STS, "tpmSts", 1);
+        }
+    }
+    fn reset(&mut self) -> Result<()> {
```
fixed

praveen-pk commented on devices/src/tpm.rs:
```diff
@@ -0,0 +1,465 @@
+        tpm.reset()?;
+        Ok(tpm)
+    }
+    fn get_active_locality(&mut self) -> u32 {
```
fixed!

likebreath commented on docs/tpm.md:
```diff
@@ -0,0 +1,122 @@
+# TPM
+tpm in Cloud-Hypervisor is emulated using `swtpm` as the backend.
```
Worth adding a link to the `swtpm`.

praveen-pk replied:
done!

likebreath reviewed (commented):
@praveen-pk Overall looking good. I have some comments left below which should be quick to adopt, besides the one about if using `phf` is worth it. I will leave it up to you to decide if you want to change as a part of this PR or not (in a follow-up PR).

Also, I was wondering if you can add an integration test for the tpm test based on document you wrote? Of course, it can be done in a separate PR.

Thank you for the hard work.

rbradford commented on tpm/src/emulator.rs:
```diff
@@ -0,0 +1,457 @@
+use crate::{TPM_CRB_BUFFER_MAX, TPM_SUCCESS};
+use anyhow::anyhow;
+use libc::c_void;
+use libc::{sockaddr_storage, socklen_t};
```
In the rest of the code base we do use `#[macro_use]` to get the macros from log. A the moment either approach is fine but we should standardise on a single way to do it in the future (but that doesn't have to be determined by this PR!)

rbradford commented:
@praveen-pk Do you want me to hold the release to integrate the feedback from @likebreath ?

praveen-pk commented:
> @praveen-pk Do you want me to hold the release to integrate the feedback from @likebreath ?

Yes please @rbradford, if thats ok. I am working on the feedback now. Will push an update in a few Hrs.

praveen-pk commented on tpm/Cargo.toml:
```diff
@@ -0,0 +1,11 @@
+[package]
+name = "tpm"
+edition = "2021"
```
done!

praveen-pk commented on tpm/src/socket.rs:
```diff
@@ -0,0 +1,139 @@
+        let size: usize = socket.read(newbuf).map_err(|e| {
+            Error::ReadFromSocket(anyhow!("Failed to read from socket. Error Code {:?}", e))
+        })?;
+        buf[0..size].clone_from_slice(&newbuf[0..size]);
```
This socket read is invoked from emulator.rs. At the time of read, the input buffer created is also of size TPM_CRB_BUFFER_MAX as is read here. Just to be safe, I return error if the input buffer is of insufficient size.

praveen-pk commented on tpm/src/emulator.rs:
```diff
@@ -0,0 +1,457 @@
+        let mut res: PtmResult = 0;
+
+        let mut fds = [-1, -1];
+        unsafe {
+            let ret = libc::socketpair(libc::AF_UNIX, libc::SOCK_STREAM, 0, fds.as_mut_ptr());
+            if ret == -1 {
+                return Err(Error::PrepareDataFd(anyhow!(
+                    "Failed to prepare data fd for tpm emulator. Error Code {:?}",
+                    std::io::Error::last_os_error()
+                )));
+            }
+        }
```
done

praveen-pk commented on tpm/src/emulator.rs:
```diff
@@ -0,0 +1,457 @@
+    /// Function to write to data socket and read the response from it
+    fn unix_tx_bufs(&mut self) -> Result<()> {
+        let isselftest: bool;
+        let mut addr: sockaddr_storage = unsafe { mem::zeroed() };
```
@likebreath could you please review the comments I added to unsafe blocks? some of these are based on SAFETY comments in the implementation of the unsafe method.

likebreath replied:
I would put: `Safe as type "sockaddr_storage" is valid with an all-zero byte-pattern value`

praveen-pk replied:
updated

praveen-pk commented on tpm/src/emulator.rs:
```diff
@@ -0,0 +1,457 @@
+            msghdr.msg_namelen = 0;
+            msghdr.msg_iov = data_vecs.as_ptr() as *mut libc::iovec;
+            msghdr.msg_iovlen = data_vecs.len() as _;
+            unsafe {
```
done

praveen-pk commented on tpm/src/emulator.rs:
```diff
@@ -0,0 +1,457 @@
+            }
+
+            cmd.output.fill(0);
+            unsafe {
```
done

praveen-pk commented on devices/src/tpm.rs:
```diff
@@ -0,0 +1,467 @@
+use arch::aarch64::layout::{TPM_SIZE, TPM_START};
+#[cfg(target_arch = "x86_64")]
+use arch::x86_64::layout::{TPM_SIZE, TPM_START};
+use phf::phf_map;
```
@likebreath I have not compared phf HashMap before choosing phf. I will revisit this after this PR gets merged. Created a bug to track this work: https://github.com/cloud-hypervisor/cloud-hypervisor/issues/4873.

praveen-pk commented on devices/src/tpm.rs:
```diff
@@ -0,0 +1,467 @@
+    (addr >> 12) as u8
+}
+
+pub struct TPM {
```
done!

praveen-pk commented on devices/src/tpm.rs:
```diff
@@ -0,0 +1,467 @@
+        let mut offset: u32 = offset as u32;
+        let data_len: usize = data.len();
+
+        if offset >= CRB_DATA_BUFFER && offset < CRB_DATA_BUFFER + (TPM_CRB_BUFFER_MAX as u32) {
```
fixed

praveen-pk commented on devices/src/tpm.rs:
```diff
@@ -0,0 +1,467 @@
+            let start: usize = (offset as usize) - (CRB_DATA_BUFFER as usize);
+            let end: usize = start + data.len();
+            let len = data.len();
+            data[0..len].clone_from_slice(&self.data_buff[start..end]);
```
done!

praveen-pk commented on devices/src/tpm.rs:
```diff
@@ -0,0 +1,467 @@
+        let locality = locality_from_addr(offset) as u32;
+        let size = data.len();
+
+        if offset >= CRB_DATA_BUFFER && offset < CRB_DATA_BUFFER + (TPM_CRB_BUFFER_MAX as u32) {
```
done! Thanks for the catch.

praveen-pk commented:
@likebreath, FYI, I created a bug to track adding integration test for TPM driver: https://github.com/cloud-hypervisor/cloud-hypervisor/issues/4872.

likebreath commented on tpm/src/emulator.rs:
```diff
@@ -0,0 +1,457 @@
+                iov_len: cmd.input.len(),
+            }; 1];
+
+            let mut msghdr: libc::msghdr = unsafe { mem::zeroed() };
```
Based on the added `safe` comments, do you intend to manually update all members of `libc::msghdr`? Assuming yes,  you are missing: `msghdr.msg_control`,`msghdr.msg_controllen`, and `msghdr.msg_flags`.

likebreath reviewed (approved):
Only have two more nits about adding "safety" comments around unsafe code. Approving the PR.  Thank you for the hard work. @praveen-pk

rbradford commented:
@praveen-pk Just a couple of nits left from @likebreath's review. Could you address those and then we can merge this.

praveen-pk commented:
> @praveen-pk Just a couple of nits left from @likebreath's review. Could you address those and then we can merge this.

@rbradford, I added all the nits from @likebreath.  Thank you!

praveen-pk commented:
@likebreath, @rbradford, @sboeuf  I appreciate all your support to get this PR merged!
